File: Medium/Python/TreesAndGraphs/BuildTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
            :type preorder: List[int]
            :type inorder: List[int]
            :rtype: TreeNode
            """
        if inorder:
            pos = inorder.index(preorder.pop(0))
            root = TreeNode(inorder[pos])
            root.left = self.buildTree(preorder, inorder[0: pos])
            root.right = self.buildTree(preorder, inorder[pos + 1:])
            return root

File: Medium/Java/TreeAndGraph/BuildTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return build(preorder, 0, inorder, 0, inorder.length);
    }
    
    public TreeNode build(int[] preorder, int ps, int[] inorder, int is, int ie) {
        if (is < ie) {
            TreeNode root = new TreeNode(preorder[ps]);
            int inIndex = 0;
            while (inorder[inIndex] != preorder[ps])
                inIndex ++;
            root.left = build(preorder, ps + 1, inorder, is, inIndex);
            root.right = build(preorder, ps + inIndex - is + 1, inorder, inIndex + 1, ie);
            return root;
        }
        return null;
    }
}
